provider/sns/broker_test.go

```diff
@@ -0,0 +1,210 @@
+package sns_test
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	awssns "github.com/aws/aws-sdk-go-v2/service/sns"
+	awssqs "github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/botchris/go-pubsub"
+	"github.com/botchris/go-pubsub/provider/sns"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSingleBroker(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	t.Run("GIVEN a sns topic and a sns broker linked to its queue", func(t *testing.T) {
+		cfg := awsConfig(t)
+		snsCli := awssns.NewFromConfig(cfg)
+		sqsCli := awssqs.NewFromConfig(cfg)
+
+		broker, err := prepareBroker(ctx, sqsCli, snsCli, "test-queue")
+		require.NoError(t, err)
+		require.NotNil(t, broker)
+
+		tRes, err := snsCli.CreateTopic(ctx, &awssns.CreateTopicInput{Name: aws.String("test-topic")})
+		require.NoError(t, err)
+		topicARN := *tRes.TopicArn
+		topic := pubsub.Topic(topicARN)
+
+		t.Run("WHEN asking for registered topics THEN one topic is informed", func(t *testing.T) {
+			topics, tErr := broker.Topics(ctx)
+
+			require.NoError(t, tErr)
+			require.Len(t, topics, 1)
+			require.EqualValues(t, topicARN, topics[0].String())
+		})
+
+		consumer1 := &consumer{}
+		consumer2 := &consumer{}
+
+		t.Run("WHEN registering two subscribers to such topic", func(t *testing.T) {
+			sub1 := pubsub.NewSubscriber(consumer1.handle)
+			require.NoError(t, broker.Subscribe(ctx, topic, sub1))
+
+			sub2 := pubsub.NewSubscriber(consumer2.handle)
+			require.NoError(t, broker.Subscribe(ctx, topic, sub2))
+
+			t.Run("THEN sns registers only one subscription as they are the same topic", func(t *testing.T) {
```

Comment:
sounds extrange to have 2 subscribers and only one subscription

Reply:
that's because here we are dealing with two "subcriber" definitions:

- logical subscriber: those defined by project, a simple function handler
- sns subscriber: which represents a destination address attached to a topic, destination address are the queues described in the diagram above; you have only one queue per broker. So all your logical subscribers are behind the same queue.